botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1084 @@
+        return result
+
+    def resolve_value(self, value, input_parameters):
+        """Given a value, check if its a function, reference or template.
+        Return the resolved value.
+
+        :type value: Any
+        :param value: A value that may or may not be a function, reference
+        or template.
+
+        :returns: A fully resolved value.
+        """
+
+        if self.is_func(value):
+            return self.call_function(value, input_parameters)
+
+        elif self.is_ref(value):
+            return input_parameters.get(value['ref'])
+
+        elif self.is_template(value):
+            return self.resolve_template_string(value, input_parameters)
+
+        return value
+
+    def evaluate_conditions(self, input_parameters):
+        """Determine if all conditions in a rule are truthy.
+
+        :returns: A bool whether or not all conditions in a rule are truthy
+        """
+        for func_signature in self.conditions:
+            if not self.call_function(func_signature, input_parameters):
+                return False
+        return True
+
+    def _is_set(self, value):
+        """**NOTE:** This is an internal API feature that is NOT
+        subject to backwards compatibility guarantees.
+
+        Evaluates whether a value (such as an endpoint parameter)
+        is set (aka not null)
+
+        :type value: Any
+        :param value: A value of any type such as an endpoint parameter
+
+        :return: A bool if the value is not None
+        """
+        return value is not None
+
+    def _get_attr(self, value, path):
+        """**NOTE:** This is an internal API feature that is NOT
+        subject to backwards compatibility guarantees.
+
+        Example: Given the input object
+
+        {"Thing1": "foo", "Thing2": ["index0", "index1"],
+        "Thing3": {"Thing3": {"SubThing": 42}}}
+
+        Thing1 returns "foo"
+        Thing2[0] returns "index0"
+        Thing3.SubThing returns 42
+        Example: Given the input array ["foo", "bar"]
+        [0] => "foo"
+
+        :type value: dict or list
+        :param value: a JSON like object that contains values, key-value pairs or
+        some combination of both
+
+        :type path: str
+        :param path: A string representing a key, index or some combination of both.
+        A period `.` indicates the value to the right of it is a nested key in a
+        dictionary, and an integer in brackets i.e. `[1]` represents an index of a list.
+        These can be used together i.e. `Thing1.foo[3]`
+
+        :returns: The object contained within the provided path
+        """
+        for part in path.split('.'):
+            match = GET_ATTR_RE.search(part)
+            if match is not None:
+                name, index = match.groups()
+                index = int(index)
+                value = value.get(name)
+                if value is None or index >= len(value):
+                    return None
+                value = value[index]
```

Comment:
The spec we are working from states that:

> An index (eg. [2]), if present, MUST, occur at the end of the path.

That means you can return at this line.

Reply:
sure i'll add a comment for clarification